coverage: Dismantle `map_data.rs` by moving its responsibilities elsewhere

Zalathar opened the pull request:
This is a series of incremental changes that combine to let us get rid of `coverageinfo/map_data.rs`, by moving all of its responsibilities into more appropriate places.

Some of the notable consequences are:

- We once again build the per-CGU file table on the fly while preparing individual covfun records, instead of building the whole table up-front. The up-front approach was introduced by #117042 to work around various other problems in generating the covmap/covfun records, but subsequent cleanups have made that approach no longer necessary.
- Expression conversion and mapping-region conversion are now performed directly in `mapgen::covfun`, which should make future changes easier.
- We no longer insert unused function instances into the same map that is also used to track used function instances. This helps to decouple the handling of used vs unused functions.

---

There should be no meaningful change to compiler output. The file table is no longer sorted, because reordering it would invalidate the file indices stored in individual covfun records, but the table order should still be deterministic (albeit arbitrary).

There are some subsequent cleanups that I intend to investigate, but this is enough change for one PR.

Zalathar committed:
coverage: Pull expression conversion out of `map_data.rs`

Zalathar committed:
coverage: Pull region conversion out of `map_data.rs`

Zalathar committed:
coverage: Build the global file table on the fly

Zalathar committed:
coverage: Track used functions in a set instead of a map

This patch dismantles what was left of `FunctionCoverage` in `map_data.rs`,
replaces `function_coverage_map` with a set, and overhauls how we prepare
covfun records for unused functions.

Zalathar committed:
coverage: Pull function source hash out of `map_data.rs`

Zalathar committed:
coverage: Use `is_eligible_for_coverage` to filter unused functions

The checks in `is_eligible_for_coverage` include `is_fn_like`, but will also
exclude various function-like things that cannot possibly have coverage
instrumentation.

Zalathar commented:
One of the big enablers for this was #134029, which made `map_data` no longer responsible for identifying unused counter/expression IDs.

jieyouxu commented:
r? jieyouxu

Zalathar commented:
I don't remember exactly why I reversed the order in the first place, but I think it had something to do with wanting to change the relative order of some other bits of code that happened to be intertwined with the construction of the file table.

Those intertwinings have been dissolved by other changes, so now there's no particular reason to avoid the more “natural” ordering of steps.

Zalathar commented:
I noticed that `IndexSet::insert_full` exists, which does precisely what I want when interacting with the global file table [(diff)](https://github.com/rust-lang/rust/compare/eb1a93ce50fae3cf6605c9f0c6230fc9f56d358a..855ebdff978df5a545cf237fb65ea55002581450).

jieyouxu commented:
@Zalathar could you elaborate on

> The file table is no longer sorted, because reordering it would invalidate the file indices stored in individual covfun records, but the table order should still be deterministic (albeit arbitrary).

So what is the deterministic order?

jieyouxu commented on compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs:
```diff
@@ -268,20 +261,17 @@ fn add_unused_functions(cx: &CodegenCx<'_, '_>) {
-        // Get the coverage info from MIR, skipping functions that were never instrumented.
-        let body = tcx.optimized_mir(def_id);
-        let Some(function_coverage_info) = body.function_coverage_info.as_deref() else { continue };
+    // FIXME(79651): Consider trying to filter out dummy instantiations of
```
Nit: `FIXME(#79651)` will make github render the issue number as a clickable link during git blame or rich source view :P

Zalathar replied:
This is existing code being reindented, but I might as well fix it anyway.

jieyouxu reviewed (changes requested):
Thanks, this LGTM overall, I just have some nits.

And also would be very cool if you could include a brief comment on the table order.

Zalathar commented on compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs:
```diff
@@ -137,22 +132,13 @@ struct GlobalFileTable {
-            bug!("file name not found in prepared global file table: {file_name}");
-        });
+    fn global_file_id_for_file_name(&mut self, file_name: Symbol) -> GlobalFileId {
+        // Ensure the given file has a table entry, and get its index.
+        let (raw_id, _) = self.raw_file_table.insert_full(file_name);
```
Yep, the intended behaviour is:
- If the file name isn't already in the table, insert it
- Now that it's guaranteed to be in the table, get its index

But the `insert` and `insert_full` methods are idempotent, so a “conditional insert” is the same thing as just doing a normal insert, and `insert_full` correctly returns the index in either case. The only catch is that I have to throw away the boolean that indicates whether the file name previously existed or not, since we don't care about that distinction.

Zalathar commented on compiler/rustc_codegen_llvm/src/coverageinfo/map_data.rs:
```diff
@@ -1,31 +1,24 @@
-        function_coverage_info: &'tcx FunctionCoverageInfo,
+        _function_coverage_info: &'tcx FunctionCoverageInfo,
         ids_info: &'tcx CoverageIdsInfo,
     ) -> Self {
```
This is temporary; the whole file gets deleted later in the same PR.

But I realise now that it's cleaner to just mark the *field* unused, and leave the parameters alone, so I'll do that instead.

jieyouxu replied:
Ah indeed

jieyouxu commented on compiler/rustc_codegen_llvm/src/coverageinfo/map_data.rs:
```diff
@@ -1,31 +1,24 @@
-        Self { function_coverage_info, ids_info: None }
+    pub(crate) fn new_unused(_function_coverage_info: &'tcx FunctionCoverageInfo) -> Self {
+        Self { ids_info: None }
     }
```
EDIT: same as the other review comment.

Zalathar commented:
Addressed review feedback [(diff)](https://github.com/rust-lang/rust/compare/855ebdff978df5a545cf237fb65ea55002581450..fc2b407c6e91f36884b0f6b2c2d1447d4010fc44).

Zalathar commented on compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs:
```diff
@@ -271,16 +271,14 @@ fn add_unused_functions(cx: &CodegenCx<'_, '_>) {
     let usage = prepare_usage_sets(tcx);
 
     let is_unused_fn = |def_id: LocalDefId| -> bool {
```
I took a slightly different approach of doing `let d: DefId = LocalDefId::to_def_id(def_id);` (with a comment), in the hope that the rest of the function can just be read intuitively without sweating the details. Let me know what you think.

jieyouxu replied:
That's looks good, as long as it's not ambiguous as it was before.

jieyouxu reviewed (approved):
Thanks. All other changes LGTM, just waiting on confirmation if we can rely on the traversal order (or whatever order) to be *deterministic*.

r=me if that is indeed the case -- that we can *rely* on the traversal order being guaranteed deterministic, unless it only *happens* to currently be that way, but is subject to changes.

Zalathar commented:
OK, even though I'm confident that the traversal order is deterministic, I ended up adding a layer of sorting (by symbol name) so that changes to stable-hashing won't influence the order of the global file table, which can be observed by coverage test snapshots [(diff)](https://github.com/rust-lang/rust/compare/9df00345038eaa2b0c5d63e595f7cf9481350ec1..541d4e85d90b4c0416014c3620d3f3c9771bd476).

(Redacting this detail in the tests themselves would be annoying, because it's embedded in the bytes of each individual covfun record.)

Zalathar commented:
@rustbot ready

jieyouxu reviewed (approved):
Thanks, the changes LGTM.

jieyouxu commented:
@bors r+

bors commented:
:pushpin: Commit 541d4e85d90b4c0416014c3620d3f3c9771bd476 has been approved by `jieyouxu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu 541d4e85d90b4c0416014c3620d3f3c9771bd476 -->
<!-- homu: {"type":"Approved","sha":"541d4e85d90b4c0416014c3620d3f3c9771bd476","approver":"jieyouxu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

marxin commented:
@Zalathar 
Can you please address the following `rustc-dev-guide` orphan links:

```
error: Server returned 404 Not Found for https://doc.rust-lang.org/nightly/nightly-rustc/rustc_codegen_llvm/coverageinfo/map_data/struct.FunctionCoverage.html
    ┌─ llvm-coverage-instrumentation.md:199:56
    │
199 │   expression ID is passed through to the corresponding [`FunctionCoverage`]
    │                                                        ^^^^^^^^^^^^^^^^^^^^ Server returned 404 Not Found for https://doc.rust-lang.org/nightly/nightly-rustc/rustc_codegen_llvm/coverageinfo/map_data/struct.FunctionCoverage.html

error: Server returned 404 Not Found for https://doc.rust-lang.org/nightly/nightly-rustc/rustc_codegen_llvm/coverageinfo/mapgen/fn.add_unused_functions.html
    ┌─ llvm-coverage-instrumentation.md:287:1
    │
287 │ [`add_unused_functions()`][add-unused-functions]:
    │ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Server returned 404 Not Found for https://doc.rust-lang.org/nightly/nightly-rustc/rustc_codegen_llvm/coverageinfo/mapgen/fn.add_unused_functions.html

error: Server returned 404 Not Found for https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir_transform/coverage/graph/struct.TraverseCoverageGraphWithLoops.html
    ┌─ llvm-coverage-instrumentation.md:447:1
    │
447 │ [`TraverseCoverageGraphWithLoops`][traverse-coverage-graph-with-loops]
    │ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Server returned 404 Not Found for https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir_transform/coverage/graph/struct.TraverseCoverageGraphWithLoops.html

error: Server returned 404 Not Found for https://doc.rust-lang.org/nightly/nightly-rustc/rustc_interface/queries/struct.Query.html
```
